### DISCUSSION TITLE
Usage with CALayer

### DISCUSSION THREAD

**Discussion by almagest21** [kingslay/KSPlayer#262] [Q&A] [answered]
Thank you for this brilliant library.
I'm developing a viewer that it has some complex CALayer based thumbnails list and detailed view for media files.
Can I use KSPlayer with CALayer like using Apple's AVPlayerLayer?
I checked Demo App but it's AudioPlayerView is based on UIView.

**Comment by kingslay** [comment]
you can use view.layer

**Reply by almagest21** [reply]
Thank you for your quick reply.
I tried below code but video output is not shown on the parent view. (I confirmed the cat.mp4 is surely shown and playable with "view.addSubview(playerView)" case.)

`

    class ViewController: UIViewController {

        override func viewDidLoad() {
            super.viewDidLoad()

            KSPlayerManager.secondPlayerType = KSMEPlayer.self
            let playerView = IOSVideoPlayerView()
            guard let contentUrl = Bundle.main.url(forResource: "cat", withExtension: "mp4") else { return }
            playerView.set(url:contentUrl,options: KSOptions())
            
            view.layer.addSublayer(playerView.layer)
            playerView.layer.frame = CGRect(x: 0, y: 0, width: 200, height: 200)
            playerView.play()
        }
    }
`

**Reply by almagest21** [reply]
Sorry, do you have any ideas? Or is there any mistakes in my usage?

**Comment by kingslay** [comment]
```
 override func viewDidLoad() {
        super.viewDidLoad()
        view.addSubview(playerView)
        playerView.delegate = self
//        playerView.translatesAutoresizingMaskIntoConstraints = false
        #if os(iOS)
        view.layer.addSublayer(playerView.layer)
        playerView.layer.frame = CGRect(x: 0, y: 0, width: 200, height: 200)
        #else
        NSLayoutConstraint.activate([
            playerView.topAnchor.constraint(equalTo: view.topAnchor),
            playerView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            playerView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            playerView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
        ])
        #endif
        view.layoutIfNeeded()
        playerView.backBlock = { [unowned self] in
            #if os(iOS)
            if UIApplication.shared.statusBarOrientation.isLandscape {
                self.playerView.updateUI(isLandscape: false)
            } else {
                self.navigationController?.popViewController(animated: true)
            }
            #else
            self.navigationController?.popViewController(animated: true)
            #endif
        }
        playerView.becomeFirstResponder()
    }
```

**Reply by almagest21** [reply]
Thank you for your code. Is addSubview still necessary?

The reason for using CALayer is to use Core Animation-based animations, Core Image filters, and to mirror a single piece of content onto multiple layers at the same time to zoom in on a portion of it. This already works very well with images.

I asked my first question because I want to insert a layer that displays the video content played by the KSPlayer into an existing CALayer hierarchy that has already been built as such.
In this use case, the player's GUI controls conflict with the existing GUI, so they would need to be displayed in a completely different location, not on top of the video. The built-in GUI of the video player, i.e., play button, seek bar, etc., needs to be hidden. What is most needed is video output.

Sorry for the  later detailed this information, but is there any way to display only the video output of the KSPlayer to just CALayer content in such a use case?

**Comment by kingslay** [comment]
is can not addSubview. i fixed by dbf6e757059afe8e74193c73f0ab4d06a59e2508

**Reply by almagest21** [reply]
Could you explain about your answer in detail?

**Comment by kingslay** [comment, marked as answer]
I have fixed this issue. You pull down the latest code. Now addSubview is no longer needed with CALayer

**Reply by almagest21** [reply]
Thank you very much for your quick fix. I confirmed video output on CALayer successfully with https://github.com/kingslay/KSPlayer/commit/dbf6e757059afe8e74193c73f0ab4d06a59e2508.

Unfortunately a small problem found: KSPlayer's original playback GUI appears for a few seconds at the beginning of the start (Also It will be hidden automatically). Can I hide these GUI from the beginning?

**Reply by kingslay** [reply]
Then you should use KSPlayerLayer. it has no playback GUI

**Reply by almagest21** [reply]
Thank you for your aswer. I tried below code and I get video output with no GUI successfully.

    let playerLayer = KSPlayerLayer()
    override func viewDidLoad() {
        super.viewDidLoad()
        guard let contentUrl = Bundle.main.url(forResource: "cat", withExtension: "mp4") else { return }
        
        KSPlayerManager.secondPlayerType = KSMEPlayer.self
        playerLayer.delegate = self

        playerLayer.frame = CGRect(x: 0, y: 0, width: 200, height: 200)
        view.layer.addSublayer(playerLayer.layer)

        playerLayer.set(url:contentUrl,options: KSOptions())
        playerLayer.play()
    }
    
    //MARK: KSPlayerLayerDelegate
    func player(layer: KSPlayerLayer, state: KSPlayerState) {
        print(state)
    }
    ...

I would like to ask this question for the last time.
Is this correct usage of KSPlayerLayer especially about initialization? (If I moved initialization code "let playerLayer = KSPlayerLayer()" in viewDidLoad function, I can't get video output.)

**Reply by kingslay** [reply]
it is because playerLayer deinit after viewDidLoad. you can 
```
    var playerLayer:  KSPlayerLayer!
    override func viewDidLoad() {
        super.viewDidLoad()
        playerLayer = KSPlayerLayer()
...
```

**Reply by almagest21** [reply]
I Understood. Thank you for everything.
